**Predicting Molecular Properties**

In the following notebook I hope to perform an inital exploratory data anaalysis of the datasests provided and represent the findings as clearly as possible

-----------------------------------------------------------------------------------------
Load in all the required dependancies

In [1]:
#!pip install missingno 
#libarary to visualise missing data 

In [2]:
import pandas as pd
import missingno as msno

Read in all the data into relevant dataframes 

In [3]:
train = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\train.csv")
structures = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\structures.csv")
scalar_coupling_contributions = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\scalar_coupling_contributions.csv")
potential_energy = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\potential_energy.csv")
mulliken_charges = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\mulliken_charges.csv")
magnetic_shielding_tensors = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\magnetic_shielding_tensors.csv")
dipole_moments = pd.read_csv("C:\\Users\\samir\\Desktop\\machineLearningPMP19\\dipole_moments.csv")

-----------------------------------------------------------------------
-----------------------------------------------------------------------
***Below we shall investigate the contence of each dataset***



-----------------------------------------------------------------------
-----------------------------------------------------------------------

**train**

-----------------------------------------------------------------------

In [4]:
train.head()

id     molecule_name  atom_index_0  atom_index_1  type  \
0   0  dsgdb9nsd_000001             1             0  1JHC   
1   1  dsgdb9nsd_000001             1             2  2JHH   
2   2  dsgdb9nsd_000001             1             3  2JHH   
3   3  dsgdb9nsd_000001             1             4  2JHH   
4   4  dsgdb9nsd_000001             2             0  1JHC   

   scalar_coupling_constant  
0                   84.8076  
1                  -11.2570  
2                  -11.2548  
3                  -11.2543  
4                   84.8074

* **id** coloumn gives the unique identitcation of the molecule and is used in the final submission format.
* **molecule_name** gives us the molecule name which is used in other datasets to refer to specific molecules.
* **atom_index_0** and **atom_index_1** are the two atoms of the molecule being considered. 
* **type** the type of coupling occuring.
* **scalar_coupling_constant** is the value we aim to predict.

In [5]:
train.index


RangeIndex(start=0, stop=4658147, step=1)

There are 4,658,147 unique coupling interactions (are they all unique?)

------------------------------------------------------------------------

**Structures**

In [6]:
structures.head()

molecule_name  atom_index atom         x         y         z
0  dsgdb9nsd_000001           0    C -0.012698  1.085804  0.008001
1  dsgdb9nsd_000001           1    H  0.002150 -0.006031  0.001976
2  dsgdb9nsd_000001           2    H  1.011731  1.463751  0.000277
3  dsgdb9nsd_000001           3    H -0.540815  1.447527 -0.876644
4  dsgdb9nsd_000001           4    H -0.523814  1.437933  0.906397

This dataset gives the position of each atom in each molecule, expressed in terms of its x,y,z position in a carteasian space

In [7]:
structures.index

RangeIndex(start=0, stop=2358657, step=1)

There are 2,358,657 atoms making up our molecules


------------------------------------------------------------------------

**Scalar coupling contributions**

In [8]:
scalar_coupling_contributions.head()

molecule_name  atom_index_0  atom_index_1  type       fc        sd  \
0  dsgdb9nsd_000001             1             0  1JHC  83.0224  0.254579   
1  dsgdb9nsd_000001             1             2  2JHH -11.0347  0.352978   
2  dsgdb9nsd_000001             1             3  2JHH -11.0325  0.352944   
3  dsgdb9nsd_000001             1             4  2JHH -11.0319  0.352934   
4  dsgdb9nsd_000001             2             0  1JHC  83.0222  0.254585   

       pso       dso  
0  1.25862  0.272010  
1  2.85839 -3.433600  
2  2.85852 -3.433870  
3  2.85855 -3.433930  
4  1.25861  0.272013

* fc (Fermi contact contribution) 
* sd (Spin dipolar contrribution)
* Pso (Paramegnetic spin-orbit contribution)
* Dso (Diamagnetic spin-orbit contribution)

**the target variable (scalar_coupling_contribution) is the sum of the four above values**

In [9]:
scalar_coupling_contributions.index

RangeIndex(start=0, stop=4658147, step=1)

There are 4,658,147 entries equal to the training dataset as we would expect (this in some ways is just a more detailed breakdown)


---------------------------------------------------------------------------------------------------------------

**Potential Energy**

In [10]:
potential_energy.head()

molecule_name  potential_energy
0  dsgdb9nsd_000001         -40.52368
1  dsgdb9nsd_000002         -56.56025
2  dsgdb9nsd_000003         -76.42608
3  dsgdb9nsd_000005         -93.42849
4  dsgdb9nsd_000007         -79.83869

This dataset gives us the overall potential energy of each molecule

In [11]:
potential_energy.index

RangeIndex(start=0, stop=85003, step=1)

From this data set we can see that there are 85,003 molecules we are working with (One would hope that they are all unique)

-------------------------------------------------------------------------------------------------------------------------

**Mulliken Charge**

In [12]:
mulliken_charges.head()

molecule_name  atom_index  mulliken_charge
0  dsgdb9nsd_000001           0        -0.535689
1  dsgdb9nsd_000001           1         0.133921
2  dsgdb9nsd_000001           2         0.133922
3  dsgdb9nsd_000001           3         0.133923
4  dsgdb9nsd_000001           4         0.133923

This dataset provides the mulliken charge of each atom in every molecule

In [13]:
mulliken_charges.index

RangeIndex(start=0, stop=1533537, step=1)

Unsure why there are 1,533,537 entires. Must investigate


----------------------------------------------------------------------------------------------

**Magnetic Shielding Tensor**

In [14]:
magnetic_shielding_tensors.head()

molecule_name  atom_index        XX      YX      ZX      XY        YY  \
0  dsgdb9nsd_000001           0  195.3150  0.0000 -0.0001  0.0000  195.3170   
1  dsgdb9nsd_000001           1   31.3410 -1.2317  4.0544 -1.2317   28.9546   
2  dsgdb9nsd_000001           2   31.5814  1.2173 -4.1474  1.2173   28.9036   
3  dsgdb9nsd_000001           3   31.5172  4.1086  1.2723  4.1088   33.9068   
4  dsgdb9nsd_000001           4   31.4029 -4.0942 -1.1793 -4.0944   34.0776   

       ZY      XZ      YZ        ZZ  
0  0.0007 -0.0001  0.0007  195.3170  
1 -1.7173  4.0546 -1.7173   34.0861  
2 -1.6036 -4.1476 -1.6036   33.8967  
3  1.6950  1.2724  1.6951   28.9579  
4  1.6259 -1.1795  1.6260   28.9013

contains the elements of the matrix making up the magnetic shielding tensor (Rank 1). Appears to be a three by three matrix

In [15]:
magnetic_shielding_tensors.index

RangeIndex(start=0, stop=1533537, step=1)

There are 1,533,537 entries, the same as the number for the mulliken charges. 


-------------------------------------------------------------------------------------------

**Dipole Moments**

In [16]:
dipole_moments.head()

molecule_name       X    Y       Z
0  dsgdb9nsd_000001  0.0000  0.0  0.0000
1  dsgdb9nsd_000002 -0.0002  0.0  1.6256
2  dsgdb9nsd_000003  0.0000  0.0 -1.8511
3  dsgdb9nsd_000005  0.0000  0.0 -2.8937
4  dsgdb9nsd_000007  0.0000  0.0  0.0000

Gives the x, y, z components of the vector which represent the overall dipole moment of the molecule

In [17]:
dipole_moments.index


RangeIndex(start=0, stop=85003, step=1)

There are 85,003 entries corresponding exactly as we would expect to the number of molecules present